Main.hs

```diff
@@ -113,100 +113,169 @@ hsQName builtins f
+  _ -> Nothing
+
+isSpecialType :: QName -> Maybe (QName -> Elims -> TCM (Hs.Type ()))
+isSpecialType = show >>> \ case
+  "Haskell.Prelude.Tuple" -> Just tupleType
+  _ -> Nothing
+
+isSpecialName :: QName -> Maybe (Hs.QName ())
+isSpecialName = show >>> \ case
+    "Agda.Builtin.Nat.Nat"         -> unqual "Integer"
+    "Agda.Builtin.Float.Float"     -> unqual "Double"
+    "Agda.Builtin.Bool.Bool.false" -> unqual "False"
+    "Agda.Builtin.Bool.Bool.true"  -> unqual "True"
+    "Agda.Builtin.List.List"       -> special Hs.ListCon
+    "Agda.Builtin.List.List._∷_"   -> special Hs.Cons
+    "Agda.Builtin.List.List.[]"    -> special Hs.ListCon
+    "Agda.Builtin.Unit.⊤"          -> special Hs.UnitCon
+    "Agda.Builtin.Unit.tt"         -> special Hs.UnitCon
+    "Haskell.Prelude.Tuple.[]"     -> special Hs.UnitCon
+    _ -> Nothing
   where
-    unqual n  = Hs.UnQual () $ hsName n
-    special c = Hs.Special () $ c ()
+    unqual n  = Just $ Hs.UnQual () $ hsName n
+    special c = Just $ Hs.Special () $ c ()
+
+ifThenElse :: QName -> Elims -> TCM (Hs.Exp ())
+ifThenElse _ es = compileArgs es >>= \case
+  -- fully applied
+  (b : t : f : es') -> return $ Hs.If () b t f `eApp` es'
+  -- partially applied -> eta-expand
+  es' -> do
+    xs <- fmap Hs.name . drop (length es') <$> mapM freshString ["b", "t", "f"]
+    let [b, t, f] = es' ++ map Hs.var xs
+    return $ Hs.lamE (Hs.pvar <$> xs) $ Hs.If () b t f
+
+tupleType :: QName -> Elims -> TCM (Hs.Type ())
+tupleType _ es | Just [as] <- allApplyElims es = do
+  err <- sep [ text "Argument"
+             , nest 2 $ prettyTCM as
+             , text "to Tuple is not a concrete list" ]
+  xs <- makeList err (unArg as)
+  ts <- mapM compileType xs
+  return $ Hs.TyTuple () Hs.Boxed ts
+tupleType _ es =
+  genericDocError =<< text "Bad tuple arguments: " <?> prettyTCM es
+
+tupleTerm :: ConHead -> ConInfo -> Elims -> TCM (Hs.Exp ())
+tupleTerm cons i es = do
+  let v = Con cons i es
+  err <- sep [ text "Tuple value"
+             , nest 2 $ prettyTCM v
+             , text "does not have a known size." ]
+  xs <- makeList' "Haskell.Prelude.Tuple.[]" "Haskell.Prelude.Tuple._∷_" err v
+  ts <- mapM compileTerm xs
+  return $ Hs.Tuple () Hs.Boxed ts
+
+tuplePat :: ConHead -> ConPatternInfo -> [NamedArg DeBruijnPattern] -> TCM (Hs.Pat ())
+tuplePat cons i ps = do
+  let p = ConP cons i ps
+  err <- sep [ text "Tuple pattern"
+             , nest 2 $ prettyTCM p
+             , text "does not have a known size." ]
+  xs <- makeListP' "Haskell.Prelude.Tuple.[]" "Haskell.Prelude.Tuple._∷_" err p
+  qs <- mapM compilePat xs
+  return $ Hs.PTuple () Hs.Boxed qs
 
 -- Compiling things -------------------------------------------------------
 
-compile :: Options -> Builtins -> IsMain -> Definition -> TCM CompiledDef
-compile _ builtins _ def = getUniqueCompilerPragma pragmaName (defName def) >>= \ case
-  Just _  -> compile' builtins def
+compile :: Options -> () -> IsMain -> Definition -> TCM CompiledDef
```

Comment:
For consistency and resilience to future additions, I would put `ModuleEnv` here (instead of `()`).